libraries/chain/fork_database.cpp

```diff
@@ -712,7 +665,34 @@ namespace eosio::chain {
    }
 
    void fork_database::close() {
-      apply<void>([&](auto& forkdb) { forkdb.close(data_dir / config::forkdb_filename); });
+      auto fork_db_file {data_dir / config::forkdb_filename};
+      bool legacy_valid  = fork_db_l && fork_db_l->has_root();
+      bool savanna_valid = fork_db_s && fork_db_s->has_root();
+
+      // check that fork_dbs are in a consistent state
+      if (!legacy_valid && !savanna_valid) {
+         elog( "fork_database is in a bad state when closing; not writing out '${filename}', legacy_valid=${l}, savanna_valid=${s}",
+               ("filename", fork_db_file)("l", legacy_valid)("s", savanna_valid) );
+         return;
+      }
+
+      std::ofstream out( fork_db_file.generic_string().c_str(), std::ios::out | std::ios::binary | std::ofstream::trunc );
```

Comment:
Seems a bit odd to use `cfile` to read and `std::ofstream` to write. Might be better to use `cfile`.

Reply:
It was that way before my change. But yes I agree it is a little bit inconsistent. I can change it if you think it is worthwhile.